### PR TITLE
If cleanup + bugfixes

### PR DESCRIPTION
**Pull request by qndel**
This has a funny side effect - visual studio 2019 doesn't get the "internal compiler error" anymore
Rewrote most of the big ifs using nice new functions from @glebm 
A bunch of bugfix comments
Some small rewrite

### CODE REVIEW

**Comment by julealgon** [comment]
> A bunch of bugfixes

Please don't do this man... document them and send separate commits/PRs.

File: Source/monster.cpp [hunk 1]
```diff
@@ -1276,20 +1276,12 @@ void NewMonsterAnim(int i, AnimStruct *anim, Direction md, AnimationDistribution
 
 bool M_Ranged(int i)
 {
-	char ai = monster[i]._mAi;
-	return ai == AI_SKELBOW || ai == AI_GOATBOW || ai == AI_SUCC || ai == AI_LAZHELP;
+	return IsAnyOf(monster[i]._mAi,AI_SKELBOW, AI_GOATBOW, AI_SUCC, AI_LAZHELP);
```
Comment (AJenbo) [inline Source/monster.cpp:1279]:
```suggestion
	return IsAnyOf(monster[i]._mAi, AI_SKELBOW, AI_GOATBOW, AI_SUCC, AI_LAZHELP);

```

File: Source/drlg_l2.cpp [hunk 2]
```diff
@@ -1923,7 +1923,7 @@ static void CreateDoorType(int nX, int nY)
 	if (predungeon[nX][nY + 1] == 68) {
 		return;
 	}
-	if (predungeon[nX][nY] == 66 || predungeon[nX][nY] == 67 || predungeon[nX][nY] == 65 || predungeon[nX][nY] == 69) {
+	if (IsAnyOf(predungeon[nX][nY], 66, 67, 65, 69)) {
```
Comment (AJenbo) [inline Source/drlg_l2.cpp:1926]:
```suggestion
	if (IsAnyOf(predungeon[nX][nY], 65, 66, 67, 69)) {

```

File: Source/drlg_l2.cpp [hunk 3]
```diff
@@ -2717,43 +2717,23 @@ static bool CreateDungeon()
...
-				predungeon[i][j] = 35;
-			}
-			if (predungeon[i][j] == 66) {
-				predungeon[i][j] = 35;
-			}
-			if (predungeon[i][j] == 69) {
-				predungeon[i][j] = 35;
-			}
-			if (predungeon[i][j] == 65) {
+			if (IsAnyOf(predungeon[i][j], 67, 66, 69, 65)) {
```
Comment (AJenbo) [inline Source/drlg_l2.cpp:2720]:
```suggestion
			if (IsAnyOf(predungeon[i][j], 65, 66, 67, 69)) {
```

**Comment by qndel** [comment]
> > A bunch of bugfixes
> 
> Please don't do this man... document them and send separate commits/PRs.

Just as comments, I'm never going back to these lol

File: Source/drlg_l2.cpp [hunk 4]
```diff
@@ -2717,43 +2717,23 @@ static bool CreateDungeon()
...
-				if (predungeon[i][j + 1] == 32) {
-					predungeon[i][j + 1] = 35;
+				for (int a = -1; a <= 1; a++) {
+					for (int b = -1; b <= 1; b++) {
+						if (a == 0 && b == 0)
+							continue;
+						if (i + a < 0 || j + b < 0)
+							continue;
+						if (i + a >= DMAXX || j + b >= DMAXY)
+							continue;
```
Comment (AJenbo) [inline Source/drlg_l2.cpp:2732]: Possibly loop over an array of Displacements instead.
```cpp
constexpr Displacement offsets[] = { { -1, -1 }, { -1, 0 }, ... };
for (Displacement offset : offsets) {
	if (predungeon[i + offset.x][j + offset.y] == 32) {
		predungeon[i + offset.x][j + offset.y] = 35;
	}
}
````

Comment (qndel) [reply Source/drlg_l2.cpp:2732]: I'd leave it for later when we have some mechanism for looping all surrounding tiles

File: Source/drlg_l3.cpp [hunk 5]
```diff
@@ -1815,16 +1801,8 @@ static bool WoodVertD(int i, int y)
...
-		}
-		if (dungeon[i][y] == 2) {
-			return true;
-		}
-		if (dungeon[i][y] == 134) {
-			return true;
-		}
-		if (dungeon[i][y] == 136) {
+
+		if (IsAnyOf(dungeon[i][y], 7, 2, 134, 136)) {
```
Comment (AJenbo) [inline Source/drlg_l3.cpp:1805]:
```suggestion
		if (IsAnyOf(dungeon[i][y], 2, 7, 134, 136)) {
```

File: Source/drlg_l3.cpp [hunk 6]
```diff
@@ -1788,22 +1788,8 @@ static bool WoodVertU(int i, int y)
...
-		}
-		if (dungeon[i][y] == 129) {
-			return true;
-		}
-		if (dungeon[i][y] == 134) {
-			return true;
-		}
-		if (dungeon[i][y] == 136) {
+
+		if (IsAnyOf(dungeon[i][y], 7, 10, 126, 129, 134, 136)) {
```
Comment (AJenbo) [inline Source/drlg_l3.cpp:1792]:
```suggestion
		if (IsAnyOf(dungeon[i][y], 7, 10, 126, 129, 134, 136)) {
```

File: Source/drlg_l3.cpp [hunk 7]
```diff
@@ -1815,16 +1801,8 @@ static bool WoodVertD(int i, int y)
...
-			return true;
-		}
-		if (dungeon[i][y] == 2) {
-			return true;
-		}
-		if (dungeon[i][y] == 134) {
-			return true;
-		}
-		if (dungeon[i][y] == 136) {
+
```
Comment (AJenbo) [inline Source/drlg_l3.cpp:1804]:
```suggestion
```

File: Source/drlg_l3.cpp [hunk 8]
```diff
@@ -1788,22 +1788,8 @@ static bool WoodVertU(int i, int y)
...
-			return true;
-		}
-		if (dungeon[i][y] == 129) {
-			return true;
-		}
-		if (dungeon[i][y] == 134) {
-			return true;
-		}
-		if (dungeon[i][y] == 136) {
+
```
Comment (AJenbo) [inline Source/drlg_l3.cpp:1791]:
```suggestion
```

File: Source/drlg_l3.cpp [hunk 9]
```diff
@@ -1863,16 +1826,7 @@ static bool WoodHorizR(int x, int j)
...
-			return true;
-		}
-		if (dungeon[x][j] == 4) {
-			return true;
-		}
-		if (dungeon[x][j] == 135) {
-			return true;
-		}
-		if (dungeon[x][j] == 137) {
+		if (IsAnyOf(dungeon[x][j], 7, 4, 135, 137)) {
```
Comment (AJenbo) [inline Source/drlg_l3.cpp:1829]:
```suggestion
		if (IsAnyOf(dungeon[x][j], 4, 7, 135, 137)) {
```

**Review by qndel — COMMENTED** [review]
lol started a review and can't close this popup XD

File: Source/drlg_l2.cpp (hunk 4 above)
Comment (AJenbo) [reply Source/drlg_l2.cpp:2732]: what you have here is pretty cryptic though with multiple loops and skip conditions.

**Review by AJenbo — APPROVED** [review]

File: Source/engine.h [hunk 10]
```diff
@@ -43,6 +43,30 @@
...
+bool IsAnyOf(const V &v, X x)
+{
+	return v == x;
+}
+
+template <typename V, typename X, typename... Xs>
+bool IsAnyOf(const V &v, X x, Xs... xs)
+{
+	return IsAnyOf(v, x) || IsAnyOf(v, xs...);
+}
```
Comment (john-parton) [inline Source/engine.h:56]: This is the most beautiful thing I have seen all day.

**Pull request merged**